src/thread.rs

```diff
@@ -0,0 +1,483 @@
+    fn set_active_thread_id(&mut self, id: ThreadId) -> ThreadId {
+        let active_thread_id = self.active_thread;
+        self.active_thread = id;
+        assert!(self.active_thread.index() < self.threads.len());
+        active_thread_id
+    }
+
+    /// Get the id of the currently active thread.
+    fn get_active_thread_id(&self) -> ThreadId {
+        self.active_thread
+    }
+
+    /// Has the given thread terminated?
+    fn has_terminated(&self, thread_id: ThreadId) -> bool {
+        self.threads[thread_id].state == ThreadState::Terminated
+    }
+
+    /// Get the borrow of the currently active thread.
+    fn active_thread_mut(&mut self) -> &mut Thread<'mir, 'tcx> {
+        &mut self.threads[self.active_thread]
+    }
+
+    /// Mark the thread as detached, which means that no other thread will try
+    /// to join it and the thread is responsible for cleaning up.
+    fn detach_thread(&mut self, id: ThreadId) {
+        self.threads[id].detached = true;
+    }
+
+    /// Mark that the active thread tries to join the thread with `joined_thread_id`.
+    fn join_thread(&mut self, joined_thread_id: ThreadId) {
+        assert!(!self.threads[joined_thread_id].detached, "Bug: trying to join a detached thread.");
+        assert_ne!(joined_thread_id, self.active_thread, "Bug: trying to join itself");
+        assert!(
+            self.threads
+                .iter()
+                .all(|thread| thread.state != ThreadState::BlockedOnJoin(joined_thread_id)),
+            "Bug: multiple threads try to join the same thread."
+        );
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_random_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
```

Comment:
Does this really use randomness? I don't see any code for that.

Reply:
Changed to `unblock_some_thread`.